### ISSUE TITLE
Set-PSFLoggingProvider to AzureLogAnalytics

### ISSUE THREAD

**Issue by C-o-s-m-i-n** [PowershellFrameworkCollective/psframework#439]
Hi, 
I'm trying to configure **logging for an Azure Function to AzureLogAnalytics** via PSFramework but it fails for some reason and I don't really know how I should go about debugging:
**$PSVersionTable :** 

```
Name                           Value
----                           -----
PSVersion                      7.0.3
PSEdition                      Core
GitCommitId                    7.0.3
OS                             Microsoft Windows 10.0.14393
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

The script I use to do this is:

```powershell
$paramSetPSFLoggingProvider = @{
    Name         = 'AzureLogAnalytics'
    InstanceName = 'MyTask'
    WorkspaceId  = '...'
    SharedKey    = '...'
    LogType      = 'PowerShell_Events'
    Enabled      = $true
}
Set-PSFLoggingProvider @paramSetPSFLoggingProvider
if (!(Get-PSFLoggingProvider -Name AzureLogAnalytics).Enabled) {
    Throw "Provider is not configured"
}
```

How could we make it work?

Have a nice day,
Cosmin

**Comment by dgoldman-msft** [comment]
Hi Cosmin. From your output did you create your AzureLogAnalytics workspace in azure? You will need the workspace id and SharedKey. These need to be inputted to the script. From there you can just run the following:

$paramSetPSFLoggingProvider = @{
    Name         = 'AzureLogAnalytics'
    InstanceName = 'MyTask'
    WorkspaceId  = '<id>'
    SharedKey    = '<key>'
    Enabled      = $true
}
Set-PSFLoggingProvider @paramSetPSFLoggingProvider

Followed by the next command: Get-PSFLoggingProvider AzureLogAnalytics -Verbose

Name              Enabled ProviderVersion IncludeModules ExcludeModules IncludeTags ExcludeTags Initialized InstallationOptional
AzureLogAnalytics True    Version_2       {}             {}             {}          {}          False       False

For more information please see: https://psframework.org/documentation/documents/psframework/logging/loggingto/azureloganalytics.html

Thanks
Dave

**Comment by C-o-s-m-i-n** [comment]
Hi Dave,

Thank you very much for the fast reply!
Indeed I am already using the **SharedKey** and **WorkspaceId** but did not add them in the issue, they are in fact in the script that I am testing. Doing the same thing works fine from my own computer both powershell 5.1 and 7.0 btw, it's just within the azure function that this fails.

Also running : Get-PSFLoggingProvider AzureLogAnalytics -Verbose
is the same as: Get-PSFLoggingProvider AzureLogAnalytics

Thanks,
Cosmin

**Comment by dgoldman-msft** [comment]
No worries. Can you clarify within the Azure Function? Are you running this from within an Azure Workspace?

Thanks
Dave

**Comment by C-o-s-m-i-n** [comment]
Hi,

I have an **Azure Workspace** and an **Azure Function** where I'm trying to post to the workspace.

![image](https://user-images.githubusercontent.com/72807650/96744745-a99c2600-13c5-11eb-99b7-6beb75c126d6.png)

![image](https://user-images.githubusercontent.com/72807650/96744551-76f22d80-13c5-11eb-93ba-f09d28c24ff6.png)

Hope this makes it a bit more clear.

Thanks,
Cosmin

**Comment by dgoldman-msft** [comment]
This does. I did it in an Azure Runbook which you need to import manually the PSFramework and then you can call the functions from there. With functions you do not have access to the modules as you would on your local machine. Take a look at this for more help:

https://docs.microsoft.com/en-us/azure/azure-functions/functions-reference-powershell?tabs=portal – Search for Dependency management

Thanks
Dave

**Comment by C-o-s-m-i-n** [comment]
Hi,

Module is correctly installed, the functions are available, dependency is
created. Functions are usable as the 'get' would have errored out otherwise.

Cosmin

**Comment by dgoldman-msft** [comment]
If that is the case you can try calling get-psfmesaage and see what is in the error stream for more information.

Thanks
Dave

On Oct 21, 2020, at 12:33 PM, Cosmin <notifications@github.com> wrote:

﻿

Hi,

Module is correctly installed, the functions are available, dependency is
created. Functions are usable as the 'get' would have errored out otherwise.

Cosmin

**Comment by C-o-s-m-i-n** [comment]
hi,

_get-psfmesaage_ and _get-psfmesaage -errors_ returns nothing

**Comment by dgoldman-msft** [comment]
Hey! You have to enable logging in azure functions and be careful so it doesn’t keep logging. You can implement the following code:

Set-PSFConfig PSFramework.Logging.Interval 500
Set-PSFConfig PSFramework.Logging.Interval.Idle 500
Start-PSFRunspace psframework.logging

At the end of the function, disable it again:
Wait-PSFMessage -Terminate

Thanks
Dave

**Comment by FriedrichWeinmann** [comment]
Heya @C-o-s-m-i-n ,
sorry about the confusion I caused here (and thanks Dave for helping out with this).
I had to disable the logging by default in Azure Functions due to cost reasons (The original iteration would keep your function open for minutes and keep causing costs, since the logging runspace just kept going).

With the settings as Dave relayed them, this should do the job for logging while minimizing additional cost incurred.
I'm closing this issue for now, feel very free to reopen or file a new issue if any questions remain unresolved.
Cheers,
Fred
